[1단계 - 음식점 목록] - 우디(박연기) 미션 제출합니다.

yeongipark opened the pull request:
### 안녕하세요, 도밥! 😄  

바닐라 자바스크립트로 컴포넌트를 분리하는 것이 처음이라 아직 많이 부족하지만, 최선을 다해 구현해보았습니다!  

부족한 부분이나 개선하면 좋을 점이 있다면 **과감하게 피드백 부탁드립니다.** 💡  
더 좋은 방향으로 나아갈 수 있도록 적극적으로 배우겠습니다!  

잘 부탁드립니다! 🙌  

## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://yeongipark.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

## 1) 이번 단계에서 가장 많이 고민했던 문제
## **1️⃣ 함수 vs 클래스**

**📌 컴포넌트를 함수로 만들까, 클래스로 만들까?**

• 처음에는 **함수형 컴포넌트와 클래스형 컴포넌트 중 어떤 방식을 사용할지 고민**하며 두 가지 방법을 모두 시도해 보았습니다.

• 하지만 컴포넌트 내에서 별도의 상태 관리가 필요하지 않았기 때문에, **클래스를 사용할 이유가 없다고 판단하여 함수형 컴포넌트로 구현**하였습니다.

## **2️⃣ 어디까지 컴포넌트로 나눌 것인가?**

**📌 컴포넌트 분리 기준**

•**반복되는 부분은 우선적으로 컴포넌트로 분리**하였습니다.

•**헤더처럼 반복되지는 않지만, 하나의 덩어리로 볼 수 있는 요소들도 컴포넌트화 하였습니다.**

•**리스트 내부의 상세 정보(이미지, 식당 이름, 설명, 거리 등)는 너무 작은 단위라서 따로 컴포넌트로 분리하지 않았습니다.**

<img width="1126" alt="스크린샷 2025-03-07 16 14 54" src="https://github.com/user-attachments/assets/f29dec6d-ab46-44b4-a5e6-5b0a9b8353fc" />


✅ 결론

• 재사용성과 유지보수성을 고려하여, **너무 세밀하게 나누기보다는 적절한 단위에서 컴포넌트를 분리**하는 것이 더 적합하다고 판단했습니다.

## **3️⃣ innerHTML vs appendChild**

**📌 DOM 조작 방법 비교**

• 처음에는 innerHTML을 사용했습니다.

• 이유는 innerHTML은 appendChild보다 **코드를 작성할 때 가독성이 높고, 전체적인 구조를 한눈에 파악하기 쉬웠기 때문입니다.**

• 하지만, 부모 컴포넌트에서 **자식 컴포넌트를 요소(Element)로 전달할 경우**, innerHTML에서는 문자열만 삽입할 수 있기 때문에 **에러가 발생**했습니다.

• 이를 해결하기 위해 appendChild 방식을 사용했습니다.

✅ 결론

- **자식 엘리먼트를 삽입할 수 있는 구조가 필요했기 때문에 appendChild를 사용**하는 방식으로 수정했습니다.

## **4️⃣ 폼에서 추가한 상태 관리 방식에 대한 고민**

**리액트처럼 상태를 관리하는 방식**

- 입력 폼에 값을 입력할 때마다 상태를 업데이트하여 관리하는 방식
- **장점**: 입력과 동시에 유효성 검사를 수행할 수 있습니다.
- **단점**: 상태를 어디서 관리해야 할지 애매하고, 불필요한 복잡성을 초래할 가능성이 있습니다.

**폼 제출 시 한 번에 데이터를 가져오는 방식 (onSubmit 활용)**

- onSubmit 이벤트를 활용하여 **폼이 제출될 때 값을 한 번에 가져오는 방식**.
- **장점**: 구현이 간단하고, 불필요한 상태 관리를 하지 않아도 됩니다.
- **단점**: 없음.

✅ **선택한 이유**

- 입력할 때마다 실시간 유효성 검사를 할 필요가 없다고 판단하여, **onSubmit을 활용하는 두 번째 방식으로 구현**하였습니다.

## **5️⃣ 모달 열고 닫기 방식 고민**

**📌 두 가지 방식 비교**

**display: none을 이용한 방식**

- **장점**: 새로운 DOM을 생성하지 않으므로 비용이 적게 듭니다.
- **단점**: **모달 외부에서 모달 DOM을 직접 조작해야 합니다.** → 유지보수성이 떨어집니다.

**모달을 삭제 후 다시 생성하는 방식**

- **장점**: 모달이 스스로 데이터를 받아서 렌더링할 수 있습니다. (외부에서 직접 조작할 필요 없음).
- **단점**: 새로운 DOM을 생성하므로 성능 비용이 발생할 가능성이 있습니다.

✅ **선택한 이유**

- **모달이 스스로 데이터를 받아서 렌더링하는 것이 더 자연스럽고 유지보수성이 높다고 판단하여 2번 방식으로 구현**했습니다.

## **6️⃣ “추가하기” 버튼을 눌렀을 때 렌더링 방식**

**📌 두 가지 방식 비교**

**상태 변경 → 기존 DOM 제거 후 다시 생성**

• **단점**: 상태와 관련 없는 요소까지 불필요하게 다시 렌더링됩니다.

**querySelector로 리스트 요소를 가져와서 직접 추가**

• **장점**: 불필요한 리렌더링을 줄일 수 있습니다.

• **단점**: **DOM 외부에서 리스트 요소를 직접 조작하는 것은 유지보수성이 떨어진다고 판단했습니다**.

✅ **선택한 이유**

- **DOM 외부에서 직접 조작하는 방식은 유지보수성이 떨어질 수 있다고 판단하여, 첫 번째 방식(상태 변경 후 렌더링)으로 구현**했습니다.

## **7️⃣ 도메인은 어디까지인가?**

- 이전 미션(자동차 경주, 로또)에서는 **상태 관리하는 부분이 도메인** 역할을 했습니다.
- 하지만, 이번 미션에서는 UI를 다루는 컴포넌트가 많다 보니, **컴포넌트 부분도 도메인이라고 볼 수 있는지 고민이 많았습니다.**
- **UI를 위한 컴포넌트와 비즈니스 로직을 분리하는 것이 중요**하다고 생각하며, **상태 관리하는 부분이 도메인에 가깝다고 판단**했습니다.

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

### 2) 이번 리뷰를 통해 논의하고 싶은 부분
1️⃣ 컴포넌트 구분 기준

저는 아직까지 컴포넌트를 분리하는 명확한 기준을 정립하지 못했습니다. 지금 제가 적용한 방식이 맞는 것 같기도 하고, 아닌 것 같기도 해서 확신이 들지 않습니다. 🤔

컴포넌트를 분리할 때 어떤 기준을 적용해야 하는지가 가장 어려운 부분 중 하나인 것 같습니다. 특히, 지금처럼 작은 단위의 컴포넌트까지 분리해야 하는지, 아니면 하나의 기능 단위로 묶어서 관리하는 것이 더 적절한지 판단하기가 쉽지 않습니다.

혹시 도밥은 컴포넌트를 어떻게 구분하고 분리하는지, 어떤 기준을 적용하는지 궁금합니다! 🙌

2️⃣ 모달창 열고 닫는 방식
이 부분은 이번 프로젝트에서 가장 고민이 많았던 부분 중 하나였습니다.

가장 간단한 방식은 CSS 클래스를 토글하는 방식이었지만, 이 방식은 2단계에서 모달을 사용하는 경우 문제가 발생할 수 있다고 생각했습니다.

특히, 모달이 동적으로 변경되는 데이터를 포함해야 할 경우, 외부에서 직접 textContent를 추가하는 방식은 모달의 역할과 무결성을 해칠 수 있다고 판단했습니다.

따라서, 돔을 지우고 다시 생성하는 방식을 선택했는데, 이는 비용이 더 들 수 있지만 실제로 두 방식을 비교했을 때 성능 차이를 체감하기 어려웠습니다. 그래서 보다 구조적으로 깔끔하고 유지보수가 쉬운 방식인 DOM을 삭제하고 다시 생성하는 방법을 적용했습니다.

이러한 고민에 대해 도밥은 어떻게 생각하는지 궁금합니다! 🧐

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

yeongipark committed:
chore: 초기 환경 세팅

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
docs(README.md): 기능 명세서 작성

yeongipark committed:
chore: 스타일 파일 및 이미지 위치 변경

yeongipark committed:
feat(index.html): 음식점 목록 페이지 템플릿 작성

yeongipark committed:
test: 음식점 목록 페이지 접속하면 헤더가 보임

yeongipark committed:
refactor: 음식점 목록 페이지 접속 테스트 코드 중복 제거

yeongipark committed:
feat: 우측 상단 버튼 누르면 모달창 보임

yeongipark committed:
test: 우측 상단 버튼 누르면 모달창 보임

yeongipark committed:
test: 음식점 정보 리스트를 본다.

yeongipark committed:
refactor(header): 헤더 컴포넌트 분리

yeongipark committed:
feat: 레스토랑 컴포넌트 추가 및 초기화 함수 구현

yeongipark committed:
feat: element만들고 클래스 이름 추가하는 유틸 함수 구현

yeongipark committed:
refactor: html 코드에서 헤드 제거

yeongipark committed:
feat: 음식점 정보 아이템 컴포넌트 분리

yeongipark committed:
test: 음식점 정보 아이템 컴포넌트 분리

yeongipark committed:
feat: 음식점 정보 리스트 컴포넌트 분리

yeongipark committed:
refactor: 음식점 리스트 컴포넌트 구조 변경 및 리스트 컨테이너 추가

yeongipark committed:
feat: 공용 모달 컴포넌트 분리

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
refactor: innerHTML -> appendChild 로 변경

yeongipark committed:
refactor: 엘리먼트 속성 추가 로직 변경

yeongipark committed:
refactor: 엘리먼트 속성 추가 로직 변경

yeongipark committed:
feat: 카테고리에서 요소 선택하는 부분 컴포넌트화

yeongipark committed:
refactor: init 함수 호출 위치 변경

yeongipark committed:
chore: export -> export default로 변경

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
feat: 모달 아이템 공통 컴포넌트 분리

yeongipark committed:
chore: 매개변수 element -> renderChild 로 이름 변경

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
feat: 모달 버튼 컨테이너 컴포넌트화

yeongipark committed:
refactor: 공통 로직 유틸로 분리

yeongipark committed:
feat: 새로운 음식점 추가 모달 구성 요소 리팩토링

yeongipark committed:
chore: 파일 위치 변경

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
feat: 버튼 컴포넌트에 클릭 이벤트 추가

yeongipark committed:
refactor: modal 생성 및 제거 방식 변경

yeongipark committed:
feat: 음식점 데이터 클래스로 상태 관리

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
feat: 모달창 help text 추가

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
feat: 음식점 추가 기능 구현

yeongipark committed:
refactor: 이미지 경로 카테고리와 매칭

yeongipark committed:
feat: 음식점 추가 했을 때, 리렌더링 할 부분 함수화

yeongipark committed:
chore: console.log 삭제

yeongipark committed:
chore: 상태관리하는 폴더명 변경

yeongipark committed:
refactor: 버튼 -> 폼으로 제출 이벤트 위치 변경

yeongipark committed:
fix: required 속성 추가 로직 변경

yeongipark committed:
fix: select 태그 기본값 설정

yeongipark committed:
test: 정상적으로 음식점 등록되지는 지 테스트

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
test: 음식점 등록시 필드값을 제대로 채우지 않으면 경고창이 뜸

yeongipark committed:
test: 모달창이 열리는지 테스트

yeongipark committed:
test: 모달창이 닫히는 지 테스트

yeongipark committed:
test: 새로고침 시 추가한 데이터가 초기화 되는지 테스트

yeongipark committed:
chore: 불필요한 테스트 파일 제거

yeongipark committed:
docs: 기능 요구사항 수정

yeongipark committed:
chore: 불필요한 주석 제거

yeongipark committed:
chore: 상수 위치 변경

- 응집도를 위해 상수를 사용하는 위치로 이동

yeongipark committed:
refactor: 에러 메시지 상수화

Co-authored-by: kaori-killer <kaori-killer@users.noreply.github.com>

yeongipark committed:
fix: 옵션 검증 로직 수정

yeongipark committed:
fix: 옵션 검증 로직 재수정

yeongipark committed:
refactor: 객체로 선언한 상수를 동결

yeongipark committed:
chore: 불필요한 상수 파일 제거

yeongipark committed:
fix: 이미지 경로 변경

yeongipark committed:
chore: 더미 데이터 내용 변경

yeongipark committed:
refactor: 클릭 이벤트 함수 선언 위치 변경

Creative-Lee commented on public/style.css:
css 속성도 각 컴포넌트나 요소별로 나누고 .js 파일과 가깝게 둬보면 어떨까요~? 
각 컴포넌트의 수정사항이 발생했을 때 유지보수가 용이하다는 장점이 있을 것 같아요.

yeongipark replied:
조아요!!

yeongipark replied:
수정 커밋 : [refactor: 하나의 파일에 작성된 css를 여러 파일로 분리하여 작성](https://github.com/woowacourse/javascript-lunch/pull/201/commits/90dc749d41540a6f6db632806b745e5352b8de78)

Creative-Lee replied:
[nit]
변경사항 확인했습니다~! 추가로 코메트 남겨보면,
- Modal 내부에 들어가는 내용물에 대한 css는 하나로 합쳐져도 괜찮을 것 같습니다. 
  - 지금은 js 파일과 1대1로 매칭된것이 혼동을 줄수도 있어보여요(RestorantModal.js에는 거의 모든 내용물을 다루는데, 왜 css파일에는 속성이 하나지? 같은 생각이 들었네요)

yeongipark replied:
> 지금은 js 파일과 1대1로 매칭된것이 혼동을 줄수도 있어보여요(RestorantModal.js에는 거의 모든 내용물을 다루는데, 왜 css파일에는 속성이 하나지? 같은 생각이 들었네요)

오 저도 같은 생각이었어요!!

1대1로 매칭되어야 하는 강박이 있었는데 굳이 그럴 필요가 없었군요!!🤔

Creative-Lee commented on src/components/restaurant/RestaurantHeader.js:
```diff
@@ -0,0 +1,28 @@
+import createElement from "../../util/createElement";
+
+export default function RestaurantHeader(text) {
```
1. 잘 뚫어놓은 text 인자가 사용되지 않고 있네요~! 
2. text 는 의미파악이 어려운 것 같아서 title 같은 네이밍은 어떨까 싶어요.
3. 현재는 Header 컴포넌트의 우측 버튼에 대한 `클릭 이벤트` 초기화를 main.js 에서 하고있는데요~!
개발자가 Header에 포함된 버튼 클릭시 동작을 수정하기위해 본 컴포넌트가 아닌, main.js 파일을 뒤져야하는건 어색한것 같아요.
이벤트 초기화도 여기서 하면 어떨까요?
(외부에서 인자로 클릭 시 실행 할 함수를 받도록 변경하면 될것 같아요)

Creative-Lee replied:
위 코멘트에 대해서 생각 & 반영하고 나서 추가로 고민해보셔요
(필수가 아니니, 시간에 따라 알아서 조정하시면 됩니다)

---

3. `RestaurantHeader` 컴포넌트에서 음식점 관련 내용을 분리하고나면 `Header`라는 공통 컴포넌트가 탄생할 것 같습니다.

우선 Header라는 컴포넌트의 요구사항을 정의해봅시다.
- 사용처에서 title을 넣어줄 수 있음.
- 사용처에서 우측 버튼에 들어갈 이미지 태그와 버튼 클릭 시 실행할 함수를 넣어줄 수 있음.
  - (더 나아가면 단순히 우측에 들어갈 요소 자체를 넣어줄수도 있음)

yeongipark replied:
>현재는 Header 컴포넌트의 우측 버튼에 대한 클릭 이벤트 초기화를 main.js 에서 하고있는데요~!
개발자가 Header에 포함된 버튼 클릭시 동작을 수정하기위해 본 컴포넌트가 아닌, main.js 파일을 뒤져야하는건 어색한것 같아요.
이벤트 초기화도 여기서 하면 어떨까요?
(외부에서 인자로 클릭 시 실행 할 함수를 받도록 변경하면 될것 같아요)

맞아요! 저도 구현하면서 이벤트 처리를 main에서 하는게 맞나? 라는 생각이 들었어요!! 

>text 는 의미파악이 어려운 것 같아서 title 같은 네이밍은 어떨까 싶어요.

동의합니당!!

yeongipark replied:
>RestaurantHeader 컴포넌트에서 음식점 관련 내용을 분리하고나면 Header라는 공통 컴포넌트가 탄생할 것 같습니다.

이 부분은 저도 납득이 되는 내용입니다!!

다만, 현재 프로그램에서는 헤더가 고정되어 있고 변경될 가능성이 없기 때문에, 지금 형태로도 충분하다고 생각합니다.
만약 프로그램의 요구사항이 확장되고, 헤더가 변경되어야 하는 상황이 온다면, 그때는 도밥의 말씀처럼 공통 컴포넌트로 분리하는 것이 더 적절할 것 같습니다.

현재 요구사항에서는 굳이 그렇게 해야 할까? 🤔
지나친 추상화가 아닐까? 라는 생각도 들었습니다!!

yeongipark replied:
수정 커밋 : [refactor: querySelector를 사용하여 DOM 요소 선택 방식을 통일](https://github.com/woowacourse/javascript-lunch/pull/201/commits/33794622b658ead797e814198fd7acf159846047)

커밋을 잘못해서 커밋 명이랑 수정 내용이 맞지 않네요... 🥲

Creative-Lee replied:
>다만, 현재 프로그램에서는 헤더가 고정되어 있고 변경될 가능성이 없기 때문에, 지금 형태로도 충분하다고 생각합니다.
만약 프로그램의 요구사항이 확장되고, 헤더가 변경되어야 하는 상황이 온다면, 그때는 도밥의 말씀처럼 공통 컴포넌트로 분리하는 것이 더 적절할 것 같습니다.

우디에 생각에 동의합니다. 
다만, 지금은 학습을 하는 단계이기 때문에 최대한 다양한 공통 컴포넌트를 설계하고 작성해보는 연습을 하는게 좋다고 생각해 드린 코멘트 였습니다.(헤더를 재사용 할 꺼리가 있는 미션은 앞으로도 아마 없지 않을까 싶기도 해요)
때문에 우디의 생각대로 하셔도 좋습니다👍

Creative-Lee commented on src/main.js:
```diff
@@ -1,23 +1,19 @@
-  }
-});
+function event() {
+  const $button = document.querySelector(".gnb__button");
```
반복되는 querySelector 함수를 작은 유틸 함수로 만들어 놓으면 중복을 줄이고 코드를 깔끔하게 정리할 수 있어요. 

``` js
const $overlay = $('.overlay')
const $resultRow = $$('.result__row')
```
js에서 `$`와 `$$`를 관용적으로 querySelector, querySelectorAll 등에 사용하곤 한답니다. (jquery의 영향인것으로 알고있어요)
(브라우저 콘솔에서도 사용 가능한데, 이스터에그같은 개념인걸로 알고있습니다 ㅋㅋ)

yeongipark replied:
querySelector도 유틸함수로 만들 생각을 못했는데 도밥 말씀처럼 하면 훨씬 간결해 질 거 같아요!! 감사합니다:))

yeongipark replied:
수정 커밋 : [refactor: querySelector를 사용하여 DOM 요소 선택 방식을 통일](https://github.com/woowacourse/javascript-lunch/pull/201/commits/33794622b658ead797e814198fd7acf159846047)

Creative-Lee commented on src/util/reset.js:
`src/util` 디렉터리에는 어떤 도메인에서도 재사용 가능한 로직을 담습니다.(대개 순수함수입니다.)
이 한줄의 라인은 그렇다기 보기 어려운 것 같아요.
이정도는 필요한곳에 직접 작성해봅시다.

yeongipark replied:
네!

yeongipark replied:
수정 커밋 : [refactor: RestaurantModal 및 RestaurantListContainer에서 데이터 관리 방식 개선 및 구독 패턴 추가](https://github.com/woowacourse/javascript-lunch/pull/201/commits/f5cb4d77d270aa4b2d5571816ca08ddb575492dc)

리렌더링이 필요한 부분만 최적화하기 위해 옵저버 패턴을 도입하였습니다.
이를 통해 불필요한 전체 리렌더링을 방지하고, 특정 변경된 부분만 효율적으로 업데이트할 수 있도록 개선했습니다!
이 과정에서 reset 함수가 더 이상 필요하지 않게 되어 삭제하였습니다!

Creative-Lee replied:
질문을 조금 남겨볼게요. 대답을 해주시면 좋겠어요. 
옵저버 패턴은 저도 그리 잘 알지 못하는데요~!
리렌더링이 필요한 부분만 최적화하기 위해 선택할 수 있는 방법은 여러가지가 있었을 것 같은데, 

- 옵저버 패턴외에 고려했던 다른 방법을 설명해주시겠어요?
  - 각 방식의 장단점도 설명해주실 수 있을까요? 
- 최종적으로 우디는 왜 옵저버 패턴을 선택했는지 적어주시면 좋을것 같아요.

yeongipark replied:
옵저버 패턴외에 고려했던 다른 방법을 설명해주시겠어요?

사실, 옵저버 패턴 외에 다른 방법을 고려하지 않았습니다..!

처음에는 Restaurant 컴포넌트에서 초기 화면에 필요한 컴포넌트들을 body에 추가하는 구조를 유지하고 싶었습니다.
여러 컴포넌트에서 개별적으로 body에 추가하는 방식이 되면, 어느 시점에 추가되는지 추적하기 어려워질 것이라고 생각했기 때문입니다!

따라서, 화면 전체를 다시 그리지 않고, 필요한 부분만 리렌더링하는 방법을 고민했습니다.
처음에는 Restaurant 컴포넌트를 전부 지우고, 상태가 변경될 때 다시 추가하는 방식으로 구현했습니다.
그러나, 이 방식은 굳이 다시 렌더링할 필요가 없는 RestaurantHeader까지 삭제되는 문제가 발생했습니다!

이런 고민을 하던 중 옵저버 패턴을 알게 되었고, 옵저버 패턴을 적용하면 상태가 변경될 때, 해당 상태를 구독하는 곳에서만 리렌더링이 발생할 수 있습니다.
즉, 불필요한 컴포넌트 리렌더링 없이 RestaurantListContainer만 업데이트할 수 있다는 점에서 적절한 해결책이라고 판단했습니다!!

그래서 다른 방법을 고려하지 않고 바로 옵저버 패턴을 선택하여 구현했습니다!

옵저버 패턴 단점
처음에는 옵저버 패턴을 적용하고 나서 “단점이 없는 것 같은데?” 라고 생각했습니다!
그러나, 모든 기술에는 트레이드오프가 있듯이, 옵저버 패턴도 단점이 존재할 것이라 생각합니다!

제가 직접 체감하지는 못했지만, 이론적으로 생각해본 옵저버 패턴의 단점은 다음과 같습니다!

1️⃣ 구독자가 많아지면 불필요한 구독자가 생길 가능성이 있음.

2️⃣ 불필요하게 상태가 변경될 가능성이 있음

3️⃣ 디버깅이 어려워질 수 있음

하지만 현재 규모에서는 옵저버 패턴의 단점이 크게 체감되지 않아요...!

Creative-Lee commented on src/components/modal/RestaurantModalButtonContainer.js:
```diff
@@ -0,0 +1,35 @@
+import createElement from "../../util/createElement";
+import Button from "../util/Button";
+import reset from "../../util/reset";
```
사용하지 import 문은 지워봅시다 
(reset을 어디서 호출해야할까에 대한 고민이었을까요~)

yeongipark replied:
이런 간단한 부분도 놓친게 많네요....

신경써서 봐주셔서 감사합니당:):)

yeongipark replied:
수정 커밋 : [chore: 불필요한 import 제거](https://github.com/woowacourse/javascript-lunch/pull/201/commits/00c0ea154b52598e7631da98ef670645254c7558), [refactor: 사용되지 않은 import 제거](https://github.com/woowacourse/javascript-lunch/pull/201/commits/87fdb120397037b1c256911cb9128dab6ece836a)

Creative-Lee commented on src/components/restaurant/Restaurant.js:
```diff
@@ -0,0 +1,17 @@
+import RestaurantHeader from "./RestaurantHeader.js";
+import RestaurantItem from "./RestaurantItem.js";
```
Item도 사용하지 않고 있네요~ 제거해봅시다.
(미사용 import 문에 대해 경고를 표시해주는 lint를 적용해보는것도 좋겠네요)

yeongipark replied:
네!!

yeongipark replied:
수정 커밋 : [chore: 불필요한 import 제거](https://github.com/woowacourse/javascript-lunch/pull/201/commits/00c0ea154b52598e7631da98ef670645254c7558), [refactor: 사용되지 않은 import 제거](https://github.com/woowacourse/javascript-lunch/pull/201/commits/87fdb120397037b1c256911cb9128dab6ece836a)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -0,0 +1,116 @@
+  });
+  const $form = createElement({
+    tag: "form",
+    classNames: ["form"],
```
form이 2개 이상이 되었을땐 구분하기 어려운 className인것 같아요.

yeongipark replied:
마자요! 수정하겠습니다~~

yeongipark replied:
수정 커밋 : [chore: form 클래스 이름 더 직관적으로 변경](https://github.com/woowacourse/javascript-lunch/pull/201/commits/16c6861c98e9e47e770244dcf55ca9bfa6914180)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -0,0 +1,116 @@
+      const data = Object.fromEntries(new FormData($form));
+      restaurantDataList.addData(data);
+      reset();
+      init();
```
1. reset은 뭐고 init은 무엇일까요? 
다른 개발자는 이 코드만 보고 어떤 동작이 일어나는지 알 수 없다고 생각해요.
더 명확한 네이밍이 필요해보입니다.

Creative-Lee replied:
--- 위 내용 개선 후 이어서 생각해보기 --- 

2. 레스토랑 1개를 추가했을 뿐인데
  - body태그 하위의 모든 DOM을 모두 삭제하고
  - 다시 모든 DOM을 그립니다.

이런 방식을 선택한 이유를 듣고싶어서 생각을 적어주시면 좋겠고,
PR 본문에서 모달의 열고 닫기 방식에 대한 고민을 적어주셨었는데, 그 고민과 위 방식의 단점들을 엮어서 한번 더 고민해 본다면, 한번 더 개선해볼 수 있을 것 같아요.

yeongipark replied:
수정 커밋 : [refactor: RestaurantModal 및 RestaurantListContainer에서 데이터 관리 방식 개선 및 구독 패턴 추가](https://github.com/woowacourse/javascript-lunch/pull/201/commits/f5cb4d77d270aa4b2d5571816ca08ddb575492dc)

리렌더링이 필요한 부분만 최적화하기 위해 옵저버 패턴을 도입하였습니다.
이를 통해 불필요한 전체 리렌더링을 방지하고, 특정 변경된 부분만 효율적으로 업데이트할 수 있도록 개선했습니다!
이 과정에서 reset 함수가 더 이상 필요하지 않게 되어 삭제하였습니다!

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -0,0 +1,116 @@
+  $form.addEventListener("submit", function handleClickAdd(event) {
+    try {
+      event.preventDefault();
+      const $form = document.querySelector(".form");
```
이미 찾은 form을 다시 찾고 있군요 👀
event 객체에서 접근 가능할것 같습니다.

yeongipark replied:
네!!

yeongipark replied:
수정 커밋 : [refactor: createRestaurantItem 함수 내 불필요한 querySelector 제거](https://github.com/woowacourse/javascript-lunch/pull/201/commits/5a2a4f1c89adddda454cb12965416fd96542b8ac)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -0,0 +1,116 @@
+    classNames: ["form"],
+  });
+
+  $form.addEventListener("submit", function handleClickAdd(event) {
```
1. 콜백함수를 전달할 때 인라인으로 기명함수를 작성하는것보다는 아예 함수로 분리해서 코드 읽는 호흡을 끊어서 가져가도 좋을것 같습니다. 

2. submit 이벤트인데, ClickAdd라는 함수명은 조금 어색하다고 생각해요~!
click보다는 submit의 의미가 들어가는게 나아보이고, 더 개선한다면 실제로 `어떤 행동` 을 하는 것인지를 함수명으로 나타내주는게 제일 좋다고 생각합니다. (e.g. createRestaurantItem)

yeongipark replied:
수정 커밋  : [refactor: 폼 submit 이벤트에 대응할 함수 분리 및 함수명 변경](https://github.com/woowacourse/javascript-lunch/pull/201/commits/322795c97b0b2802ad272c08aa537aaa85c0bb40)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -0,0 +1,116 @@
+        Select({
+          name: "category",
+          id: "category",
+          options: ["한식", "중식", "일식", "양식", "아시안", "기타"],
```
옵션들은 상수화해서 관리하면 어떨까요?

yeongipark replied:
name과 id 같은 속성은 제외하고, options만 상수화하는 이유가 있을까요?

제가 상수화하는 기준은 다음과 같습니다:
- 상수로 표현했을 때 더 직관적인 경우
- 한 번 이상 반복해서 사용되는 경우

하지만 options의 경우, 상수화한 이름이 배열 자체를 표현하는 것과 큰 차이가 없으며,
다른 곳에서 재사용되지 않기 때문에 굳이 상수로 분리할 필요성을 느끼지 않았습니다!!

Creative-Lee replied:
기준이 명확하다면 분리하지 않아도 좋습니다.
(step2 과정에서 도메인 로직에 대해 ts를 도입하게 될텐데 그때 다시 생각해봐도 좋을 것 같네요)

Creative-Lee commented on src/main.js:
```diff
@@ -1,23 +1,19 @@
-  const buttonImage = document.createElement("img");
-  buttonImage.src = image;
+export function init() {
+  Restaurant();
+  event();
```
Restaurant() , event() 라는 단어 자체에서는 어떤 의미도 찾기 어려워 보입니다. 
함수명으로 얻을 수 있는 정보가 더 많아야하지 않을까요?

yeongipark replied:
네 동의합니다!!

yeongipark replied:
수정 커밋 : [chore: Restaurant -> RestaurantApp 으로 컴포넌트명 변경](https://github.com/woowacourse/javascript-lunch/pull/201/commits/d057570a3d5a9862c05b13ba61e4baf643327482)

Creative-Lee commented on src/components/restaurant/Restaurant.js:
```diff
@@ -0,0 +1,17 @@
+
+import restaurantDataList from "../../domain/RestaurantDataList.js";
+
+export default function Restaurant() {
```
함수명을 대문자로 시작한것은 React의 컴포넌트 관련한 룰에서 착안한 개념이 맞을까요~?
main.js 파일에도 남겨놓았지만, 
이 프로젝트는 JS로 구현되었는데, `Restaurant()` 라는 호출을 보고 어떤 동작을 하는지 유추가 가능할까요?

yeongipark replied:
> 함수명을 대문자로 시작한것은 React의 컴포넌트 관련한 룰에서 착안한 개념이 맞을까요~?

네! 컴포넌트인 함수는 컴포넌트임을 구분하기 위해 파스칼 케이스를 사용했습니다! 

도밥의 말씀처럼 Restaurant이라는 컴포넌트는 이름만으로 어떤 동작을 하는지 알 수 없을 거 같다고 생각해요!

Creative-Lee commented on src/domain/RestaurantData.js:
```diff
@@ -0,0 +1,68 @@
+    this.validateDistance(distance);
+    this.validateName(name);
+
+    this.#id = id && crypto.randomUUID();
```
로직에 오류가 있는것 같네요
의도하신 바가 맞을까요?

yeongipark replied:
헉 맞네요...! `||`로 제가 의도한대로 동작할 거 같아요!

yeongipark replied:
수정 커밋 : [fix: id 계산 로직 수정](https://github.com/woowacourse/javascript-lunch/pull/201/commits/1b51191450d3b7876ad0db143fc396f26d587300)

Creative-Lee commented on src/domain/RestaurantData.js:
```diff
@@ -0,0 +1,68 @@
+    if (this.isValidateOption(distance)) throw Error(ERROR_MASSAGE.distance);
+  }
+
+  validateName(name) {
```
url에 대한 검증도 필요해보입니다.

yeongipark replied:
네!

yeongipark replied:
수정 커밋 : [feat: 링크 유효성 검사 추가](https://github.com/woowacourse/javascript-lunch/pull/201/commits/f4c621105cac5860f3c886169ed47b3400c4155f)

Creative-Lee commented on src/domain/RestaurantData.js:
domain/ 하위 각 파일에 대해서 생각하셨던 역할을 설명해주시겠어요~?

yeongipark replied:
### RestaurantData
- 개별 음식점 정보를 저장하는 클래스입니다.
- 인스턴스가 생성될 때, 음식점 정보에 대한 유효성 검사를 수행합니다.

### RestaurantDataList
- RestaurantData 인스턴스들의 배열을 관리하는 클래스입니다.
- 음식점 정보를 추가하고, 반환하며, 구독 기능을 제공합니다.
- 구독 기능을 통해 RestaurantDataList의 상태가 변경될 때마다, 이를 구독하는 곳에서 자동으로 렌더링을 수행할 수 있습니다.

Creative-Lee commented on src/domain/RestaurantDataList.js:
```diff
@@ -0,0 +1,54 @@
+  },
+];
+
+const restaurantDataList = new RestaurantDataList(dummy);
+export default restaurantDataList;
```
요런식의 구조로 객체를 내보낸 이유가 있으실까요~?

yeongipark replied:
처음에는 RestaurantDataList의 인스턴스를 main.js에서 생성한 후, 필요한 컴포넌트나 모듈에 인수로 전달하는 방식으로 구현하려고 했습니다.

하지만 이 방식은 자식 컴포넌트가 깊이 중첩된 경우, 계속해서 인수를 전달해야 하는 문제가 발생합니다.
그렇게 되면 코드가 불필요하게 복잡해지고, 가독성이 떨어질 수 있다고 판단했습니다.

이 문제를 해결하기 위해, RestaurantDataList를 선언한 곳에서 직접 인스턴스를 생성하고 이를 export하는 방식을 선택했습니다.
이렇게 하면, RestaurantDataList의 인스턴스가 필요한 곳에서만 직접 import하여 사용하면 되므로, 불필요한 인수 전달 없이 더 깔끔한 코드 유지가 가능합니다.

이러한 이유로, 최종적으로 인스턴스를 전역적으로 생성하고 export하는 방식으로 구현하게 되었습니다!

Creative-Lee replied:
깊게 중첩된 컴포넌트에 인자를 전달하는것의 단점은 적어주셨는데, 
전역적으로 사용했을 때의 단점은 보이지 않네요.

각 방식에는 장단점이 있다고 생각하는데요
선택이 필요할때는 항상 장단을 모두 고려해서 선택해주시면 좋겠습니다~!

yeongipark replied:
전역적으로 사용했을 때의 단점은 사용하는 곳이 많아지면, 어디서 상태를 사용하고 수정하는지 파악하기 어렵다는 점이 있는 거 같아요!!

>선택이 필요할때는 항상 장단을 모두 고려해서 선택해주시면 좋겠습니다~!

네!!!

혹시 도밥의 경우 해당 미션을 해결한다면 어떤 방식으로 상태를 관리하실지 궁금해요!!!!

Creative-Lee commented on cypress/e2e/modal.cy.js:
```diff
@@ -0,0 +1,18 @@
+  it("모달에서 취소버튼을 누르면 모달창이 닫힌다.", () => {
+    cy.get(".gnb__button").click();
+
+    cy.get(".button--secondary").click();
```
button--secondary 요소가 취소 버튼임을 추측해야했어요.
class명이 변경되면 가독성이 좋아질것 같네요

yeongipark replied:
네!!

yeongipark replied:
수정 커밋 : [test: 클래스명 직관적이게 변경 + 에러 메시지 정적으로 선언](https://github.com/woowacourse/javascript-lunch/pull/201/commits/16a3fbd61f5c651855d088762c76215cbe8f091d)

Creative-Lee commented on cypress/e2e/modal.cy.js:
```diff
@@ -0,0 +1,18 @@
+  });
+
+  it("모달에서 취소버튼을 누르면 모달창이 닫힌다.", () => {
+    cy.get(".gnb__button").click();
+
```
모달이 실제로 열렸음을 확인하고 닫았음을 확인해야 조금 더 신뢰할 수 있을 것 같아요.

yeongipark replied:
바로 위에서 모달이 열리는 테스트를 진행했는데, 여기서 또 확인하는 것은 중복된 검사라고 생각하여 제외했습니다!

도밥은 그럼에도 불구하고 확인이 필요하다고 생각하시는 건가요?

Creative-Lee replied:
각각의 테스트(it 구문 하위겠죠)는 항상 독립된 단위로 생각해야합니다.
(beforeEach 구문이 각 테스트마다 실행되는 것 처럼요)
이러한 관점에서 바라보았을 때 바로 위 테스트에서 잘 열리는걸 확인했다고 해서, 다음 테스트에서도 잘 열렸다는것을 보장하는 것은 아닙니다.
때문에 `cy.visit("http://localhost:5173");`로 새롭게 방문한 곳에서는 모달의 동작을 한번 더 확인해줘야할 것 같다고 남긴거였어요.
(그런 의미에서 테스트가 합쳐져도 될것같네요)

Creative-Lee commented on cypress/e2e/registerRestaurant.cy.js:
```diff
@@ -0,0 +1,65 @@
+
+  cy.get(".button--primary").click();
+  cy.get(".modal-container").should("not.exist");
+  cy.get("body").contains("더휴");
```
리스트에 추가되었음을 확인하는게 조금 더 정확할 것 같아요.

yeongipark replied:
맞자요..😅

쉽게 하려다 보니, 위 방식으로 진행했었네요 🥲

수정하겠습니당

yeongipark replied:
수정 커밋 : [test: 음식점 추가 테스트 구체화](https://github.com/woowacourse/javascript-lunch/pull/201/commits/7ccc5fad66445ea0e095bd1b4371af5597ae3191)

Creative-Lee commented on cypress/e2e/registerRestaurant.cy.js:
```diff
@@ -0,0 +1,65 @@
+    });
+  });
+
+  it("두 글자 이하를 입력했을 때 alert 창으로 경고 문구가 발생한다.", () => {
```
**어떤 것**을 두 글자 이하 입력해야 경고가 나올까요?

yeongipark replied:
음식점 이름입니다..!

추가로 한글이나 영어 숫자만 허용하도록 해야할 거 같아요!

yeongipark replied:
수정 커밋 : [test: 테스트 이름 명확하게 작성](https://github.com/woowacourse/javascript-lunch/pull/201/commits/4dd07353ceefa072771fc5232c8093f50475ae41)

Creative-Lee commented on cypress/e2e/registerRestaurant.cy.js:
e2e 테스트에 대한 기준은 어떤것이였는지, 들려주시면 좋겠습니다~

yeongipark replied:
사실… 특별한 기준을 가지고 작성한 것은 아니었습니다! 😅
저는 E2E 테스트를 사용자의 입력 흐름을 검증하는 과정이라고 생각했습니다.
따라서, 현재 프로그램에서 사용자가 어떤 입력을 했을 때 기대하는 결과가 나오는지를 중심으로 테스트를 진행했습니다.

예를 들어:
1️⃣ 사용자가 웹사이트에 접속하면 → 헤더와 음식점 리스트가 보인다.
2️⃣ 추가하기 버튼을 클릭하면 → 모달 창이 열린다.
3️⃣ 음식점 정보를 입력하고 추가 버튼을 누르면 → 음식점이 리스트에 등록된다.

이처럼 사용자에게 발생할 수 있는 시나리오를 고려하여 E2E 테스트를 작성했습니다.
하지만 어디까지가 E2E인지, 이렇게 작성하는 것이 맞는지 아직 확신이 없습니다. 🤔

사실, 제가 생각하는 e2e는 모달창을 열어서 정상적으로 음식점이 등록됐는지 확인하는 테스트만 있으면 되겠다고 생각했었어요🧐

Creative-Lee replied:
기준은 앞으로 테스트를 작성해 나가며 세워질 부분이라고 생각합니다 😄 
저는 e2e 테스트의 범위를 생각할때 `시작부터 끝까지` 라는 큰 범위에서 생각해보는 것 같아요.
이 서비스에서 핵심이 되는 사용자 플로우가 무엇인지 생각해보면, 
'메인 페이지에서 음식점 등록 모달을 열어 음식점을 등록하고, 성공적으로 추가되어 메인페이지에 표시되는것' 인것 같고, 
이 플로우를 하나의 테스트로 작성했을 것 같네요.

정답은 없습니다. 우디가 생각했을 때, 의미있는 단위로 작성해보시면 되겠습니다~!

Creative-Lee reviewed (changes requested):
안녕하세요 우디(토이스토리 우디가 맞나요? ㅋㅋ) 만나서 반갑습니다 👋  
점심뭐먹지 미션 리뷰어 도밥입니다 잘 부탁드려요😄 


--- 
우선 가장 많이 고민하셨던 부분에서 몇가지 의견드리자면, 

> 📌 컴포넌트를 함수로 만들까, 클래스로 만들까?
> 처음에는 함수형 컴포넌트와 클래스형 컴포넌트 중 어떤 방식을 사용할지 고민하며 두 가지 방법을 모두 시도해 보았습니다.
> 컴포넌트 내에서 별도의 상태 관리가 필요하지 않았기 때문에, 클래스를 사용할 이유가 없다고 판단하여 함수형 컴포넌트로 구현하였습니다.

js로 구현하는 프로젝트에서 함수형~ 클래스형~ 이라고 표현하니, 어딘가 React의 논제같이 들리네요😅  '함수로 컴포넌트를 표현해 보았다' 정도로 받아들이겠습니다~
우디의 선택 기준에 대해서는 어느정도 납득되었습니다만, 상태관리가 필요해지면 클래스를 사용하실것인지도 궁금하네요~!

> 📌 컴포넌트 분리 기준
> ✅ 결론
재사용성과 유지보수성을 고려하여, 너무 세밀하게 나누기보다는 적절한 단위에서 컴포넌트를 분리하는 것이 더 적합하다고 판단했습니다.

컴포넌트 분리에 대해서는 훌륭한 결론을 내주셨다고 생각합니다~👍 
다만, 일부 코드에서 일관성이 조금 떨어지는 것 같아 기준에 대해 코멘트 남겨드렸으니 참고해주세요!

> 📌 폼에서 추가한 상태 관리 방식에 대한 고민
✅ 선택한 이유
입력할 때마다 실시간 유효성 검사를 할 필요가 없다고 판단하여, onSubmit을 활용하는 두 번째 방식으로 구현하였습니다.

선택한 이유에 대해서는 충분히 납득 가능했습니다.
다만, 나머지 부분은 조금 의아했는데요.
`값을 입력할 때 마다 상태를 업데이트해서 관리하는 방식`에 대해 `리액트처럼 상태를 관리하는 방식`이라는 문장이 적절한 설명인지 모르겠습니다. 
리액트에서는 폼 제출 시 한번에 데이터를 가져오는 방식을 사용하지않나요? 👀 
단점 부분도 설득력 있지는 않은 것 같아요.('관리가 애매하다'는  보는 사람에 따라 느끼는 바가 다른 주관적인 의견 같아서요)

>📌 “추가하기” 버튼을 눌렀을 때 렌더링 방식

이 역시도 장단 비교와 이유는 납득 가능했지만... 실제로 구현은 조금 다르게 된것으로 보입니다.
`기존 DOM을 제거 후 다시 생성하는 행위`의 범위가 음식점 목록 정도여야 납득이 가능할것 같은데, 현재는 body 하위에 모든 DOM을 제거 및 재생성 하고있어요.(단점이 극대화 되었다고 볼 수 있습니다.) 

>📌 도메인은 어디까지인가?
이전 미션(자동차 경주, 로또)에서는 상태 관리하는 부분이 도메인 역할을 했습니다.
하지만, 이번 미션에서는 UI를 다루는 컴포넌트가 많다 보니, 컴포넌트 부분도 도메인이라고 볼 수 있는지 고민이 많았습니다.
UI를 위한 컴포넌트와 비즈니스 로직을 분리하는 것이 중요하다고 생각하며, 상태 관리하는 부분이 도메인에 가깝다고 판단했습니다.

좋은 고민을 해주셨다고 생각합니다.
이제 저희는 이전 미션과는 다르게 컴포넌트 단위로 개발하기 시작했습니다.
컴포넌트가 view와 관련된 로직을 다룰 수도 있고, domain과 관련된 로직을 다룰 수도 있습니다.
혹은 둘중 하나만 가지고 있을수도 있죠. 때문에 컴포넌트가 도메인이냐 view이냐 나누기보다는 
오히려 우디가 생각한것처럼 `UI만 포함한 컴포넌트네?`, `이 컴포넌트는 로직을 가져와서 사용하네?` 등의 접근을 하는게 좋아보입니다~!

---

우디가 논의 하고 싶었던 부분도 봐보면,

> 1️⃣ 컴포넌트 구분 기준
컴포넌트를 분리할 때 어떤 기준을 적용해야 하는지가 가장 어려운 부분 중 하나인 것 같습니다. 특히, 지금처럼 작은 단위의 컴포넌트까지 분리해야 하는지, 아니면 하나의 기능 단위로 묶어서 관리하는 것이 더 적절한지 판단하기가 쉽지 않습니다.
혹시 도밥은 컴포넌트를 어떻게 구분하고 분리하는지, 어떤 기준을 적용하는지 궁금합니다! 🙌

일단 도메인에 종속적인 컴포넌트와 그렇지 않은 컴포넌트를 먼저 구분하면 좋은것 같아요.
이후 일단 기능을 쭉 개발합니다.
그러다 문득 `이 컴포넌트의 책임이 무엇이지?` 등의 고민을 한 뒤, 하나의 컴포넌트가 과한 책임을 가진다고 생각하면, 컴포넌트를 분리하여 책임을 나눕니다. 
과정에서는 `반드시 이 로직을 이 컴포넌트가 가져야하나?`, `자식대신 부모가 가지면 코드 흐름이 어떻게 바뀌지?` 등을 생각하는것 같습니다.  
더 나아가면 `이 로직은 다른 컴포넌트에서도 사용해야하네, 어떻게 전달하지?` 등의 고민도 하고요.

작은 단위의 컴포넌트가 어떤것을 의미하는지 명확히 모르겠지만, 현 `src/components/util` 하위에 있는 컴포넌트들 처럼 도메인과 관련없이 재사용하기 쉬운 공통 컴포넌트를 분리하는 것은 좋은 시도라고 생각합니다.


> 2️⃣ 모달창 열고 닫는 방식

위 추가하기 버튼에 대한 코멘트와 같은 답변을 드릴 수 있을 것 같습니다.
코드에도 관련한 코멘트를 남겨놨으니 한번 생각해보시면 좋겠습니다. 
(랜더링 비용에 대한 부분은 당장에 체감하긴 어렵더라도, app이 커질수록 훨씬 더 영향이 커질것은 분명합니다.) 


미션 고생하셨습니다 우디 💪 
코드 흐름과 일관성 대한 부분을 위주로 개선해보셨으면 합니다.
우선 코멘트 남긴부분들 차근히 살펴보시고, 재요청 주세요 👍

yeongipark committed:
refactor: 하나의 파일에 작성된 css를 여러 파일로 분리하여 작성

- 사용되는 컴포넌트와 가까이 놔둠으로써 컴포넌트의 스타일을 쉽게 찾을 수 있음

yeongipark committed:
refactor: querySelector를 사용하여 DOM 요소 선택 방식을 통일

- body는 document.body로 대체

yeongipark committed:
refactor: 사용되지 않은 import 제거

yeongipark committed:
chore: form 클래스 이름 더 직관적으로 변경

yeongipark committed:
refactor: 폼 submit 이벤트에 대응할 함수 분리 및 함수명 변경

yeongipark committed:
refactor: createRestaurantItem 함수 내 불필요한 querySelector 제거

yeongipark committed:
refactor: select에 들어가는 option 데이터 상수화

yeongipark committed:
fix: id 계산 로직 수정

yeongipark committed:
refactor: 음식점 아이템에 id 값 추가

yeongipark committed:
refactor: 에러 메시지 상수 위치 사용하는 컴포넌트와 가깝게 위치하도록 이동

yeongipark committed:
refactor: createElement에서 textContent와 onClick 매개변수도 추가하여 생성하도록 수정

yeongipark committed:
refactor: RestaurantModal 및 RestaurantListContainer에서 데이터 관리 방식 개선 및 구독 패턴 추가

- 기존 body 값을 모두 지웠다가 새로 렌더링하는 방식에서 옵저버 패턴을 사용해 아이템 리스트만 리렌더링 되도록 수정

- init과 reset 함수 삭제

yeongipark committed:
chore: Restaurant -> RestaurantApp 으로 컴포넌트명 변경

yeongipark committed:
chore: 불필요한 import 제거

yeongipark committed:
feat: 링크 유효성 검사 추가

yeongipark committed:
refactor: 식당 이름 유효성 검사 로직 추가

yeongipark committed:
test: 클래스명 직관적이게 변경 + 에러 메시지 정적으로 선언

- 에러 메시지를 정적으로 선언하여 다른 곳에서도 사용할 수 있도록 변경

yeongipark committed:
test: 음식점 추가 테스트 구체화

- 기존에 불명확했던 테스트를 구체화하여 테스트를 추가

yeongipark committed:
test: 테스트 이름 명확하게 작성

yeongipark committed:
test: 유효성 검사 테스트 추가

- 유효하지 않은 링크 입력 유효성 검사
- 특수 문자 입력 유효성 검사

yeongipark commented:
안녕하세요, 도밥! 🌟

(저는 토이스토리 우디 맞아요~! 🤠✨)

첫 번째 리뷰에서 너무 정성스럽게 피드백을 주셔서 감동받았습니다… 🥲
제가 미처 생각하지 못했던 부분까지 세세하게 코멘트를 남겨주셔서 정말 감사드립니다! 😄

---
# 도밥의 코멘트에 대한 🤠 생각 

> , 상태관리가 필요해지면 클래스를 사용하실것인지도 궁금하네요~!

함수로 구현할 거 같아요! 최근에 클로저에 대해 배웠는데 클로저를 사용하면 클래스를 사용하지 않고도 상태 관리를 할 수 있다고 생각이 들었어요! 그래서 상태 관리가 필요한 경우에도 함수로 구현할 거 같아요!!

> 값을 입력할 때 마다 상태를 업데이트해서 관리하는 방식에 대해 리액트처럼 상태를 관리하는 방식이라는 문장이 적절한 설명인지 모르겠습니다.
리액트에서는 폼 제출 시 한번에 데이터를 가져오는 방식을 사용하지않나요? 👀
단점 부분도 설득력 있지는 않은 것 같아요.('관리가 애매하다'는 보는 사람에 따라 느끼는 바가 다른 주관적인 의견 같아서요)

이 부분은 저 스스로 작성해놓고도 이해하기 어려운 글이네요… 😅

제가 의도했던 것은, 입력된 값을 React의 state처럼 관리하고, 이를 별도의 공간에서 관리하면서 상태가 변경될 때마다 유효성 검사를 수행하는 방식을 고민했다는 점입니다!
(머릿속에 있는 생각을 글로 표현하려니 쉽지 않네요… 😔)

하지만, 이 고민은 폼 내부의 input, select, textarea 값을 한 번에 가져올 수 있다는 사실을 몰랐을 때 했던 고민이었습니다.
모든 값을 한 번에 가져올 수 있다는 것을 알고 난 이후로는 더 이상 고민할 필요가 없었습니다! 

>기존 DOM을 제거 후 다시 생성하는 행위의 범위가 음식점 목록 정도여야 납득이 가능할것 같은데, 현재는 body 하위에 모든 DOM을 제거 및 재생성 하고있어요.(단점이 극대화 되었다고 볼 수 있습니다.)

이 부분은 부족한 시간 때문에 더 효율적인 방법이 어떤게 있는지 찾아보지 않고, 떠오르는 방식으로 바로 구현하다보니 body 하위 값들을 모두 지우는 방식으로 구현했습니다...
현재는 옵저버 패턴을 이용하여 렌더링이 필요한 곳에서만 렌더링이 일어나도록 변경했습니다!

>위 추가하기 버튼에 대한 코멘트와 같은 답변을 드릴 수 있을 것 같습니다.
코드에도 관련한 코멘트를 남겨놨으니 한번 생각해보시면 좋겠습니다.
(랜더링 비용에 대한 부분은 당장에 체감하긴 어렵더라도, app이 커질수록 훨씬 더 영향이 커질것은 분명합니다.)

앱의 규모가 커질수록 성능에 미치는 영향이 커질 것이라는 점에는 동의합니다!
그럼에도 불구하고, 이번에 모달 창을 DOM에서 제거하는 방식으로 구현한 이유는 step2에서의 모달 사용 방식 때문입니다!

step2에서는 음식점 아이템을 클릭하면 해당 음식점의 상세 정보를 담은 모달 창이 열리는 구조입니다.
만약 display: none; 방식으로 모달을 숨긴다면, HTML에 모달이 항상 존재한 상태에서 외부에서 데이터를 강제로 삽입합니다. 그러나, 모달 외부에서 직접 데이터를 조작하는 것은 구조적으로 적절하지 않다고 판단했습니다!

이에 따라, 모달이 열릴 때마다 새로운 DOM을 생성하고, 필요한 데이터를 인자로 전달하는 방식으로 구현했습니다!
이 방식은 모달 자체가 독립적으로 데이터를 관리할 수 있도록 하며,
외부에서 불필요하게 모달의 데이터를 직접 조작하지 않도록 했습니다!

성능 최적화에 대한 고민도 있었지만, **“최적화는 문제가 발생했을 때 진행해도 늦지 않다”**는 원칙(?)을 적용했습니다.
현재 규모에서는 DOM을 삭제하고 다시 추가하는 방식이 성능에 큰 부담이 되지 않는다고 판단하여, 해당 방식으로 구현하였습니다!!

따라서, 모달을 DOM에서 제거하는 방식은 단순히 UI를 숨기는 것이 아니라, 구조적으로 더 적절한 데이터 흐름을 만들기 위한 선택이었습니다!!

---

# 주요 변경 사항
### 1️⃣ 모달 백그라운 클릭, esc 클릭시 모달창 닫히는 기능 추가
이 과정에서 keydown 이벤트를 감지하기 위해 document.addEventListener("keydown", callback)을 사용했습니다!

하지만, 이벤트를 추가하기만 하면 모달 창이 열려 있지 않은 상황에서도 keydown 이벤트가 계속해서 발생하는 문제가 있었습니다.
이를 해결하기 위해 모달 창을 닫을 때, keydown 이벤트 리스너도 함께 제거하는 방식을 선택했습니다.

📌 이를 위해 모달을 닫는 함수를 만들고, 해당 함수에서 modal.remove()와 함께 이벤트 제거 코드도 추가했습니다.

🤔 도밥은 이런 방식에 대해 어떻게 생각하시나요?
혹시 더 좋은 방법이 있다면 추천해 주실 수 있을까요? 😊

---

# 질문
### 1️⃣ 상수화 위치
1번
```jsx

const CATEGORY_OPTIONS = ["한식", "중식", "일식", "양식", "아시안", "기타"];
const DISTANCE_OPTIONS = ["5분 내", "10분 내", "15분 내", "20분 내", "30분 내"];

export default function RestaurantModal() {
```

2번
```jsx
export default function RestaurantModal() {
 const CATEGORY_OPTIONS = ["한식", "중식", "일식", "양식", "아시안", "기타"];
 const DISTANCE_OPTIONS = ["5분 내", "10분 내", "15분 내", "20분 내", "30분 내"];
```
두 가지 방식이 가능했지만, 저는 2번 방식을 선택했습니다!

그 이유는, 해당 데이터는 RestaurantModal 내부에서만 필요한 값이므로,
함수 외부에 선언하면 RestaurantModal이 호출되기 전부터 이미 생성되어 있어 불필요하게 메모리를 차지할 것이라고 생각했기 때문입니다!

다만, 함수 내부에 선언하면 코드가 길어지고, 가독성을 해치는 요소가 될 수도 있다고 느꼈습니다.
어떤 면에서는 RestaurantModal의 코드를 이해하는 데 방해가 될 수도 있고, 코드가 직관적으로 보이지 않는다는 점이 고민이었습니다!

혹시 도밥은 이런 경우에 함수 내부 선언과 외부 선언 중 어떤 방식을 선호하는지, 그리고 그 이유가 궁금합니다! 😊

### 2️⃣ 선언형 방식
다른 크루의 PR을 보던 중 선언형 방식에 대해 알게 되었습니다!
선언형 프로그래밍은 ‘어떻게’ 보다는 ‘무엇’을 원하는지 선언하는 방식이라고 배웠는데,
예시를 봐도 아직 정확한 개념이 와닿지 않고, 어떻게 적용해야 할지 고민이 됩니다. 🥲

하지만 선언형 방식이 가독성을 높이고, 코드가 더 간결해진다는 점은 분명히 느껴졌습니다!
그래서 현재 코드를 선언형 방식으로 리팩토링하고 싶지만,
어떤 부분이 선언형이고, 어떤 부분이 명령형인지 명확히 구분하는 게 어렵습니다.

혹시 도밥께서 선언형 프로그래밍을 구현할 때의 꿀팁이 있다면 공유해 주세요! 😊
또한, 제 코드에서 선언형과 명령형의 예시를 하나씩 들어주시면 선언형 프로그래밍을 이해하는 데 큰 도움이 될 것 같습니다! 🚀

### 3️⃣ 콜백 함수 선언 위치
addEventListener내부에 넘겨줄 콜백 함수를 분리하다가, 이를 현재 addEventListener를 호출하는 함수 내부에 함수를 선언할 지, 아니면 외부에 선언할 지 고민했습니다. 

내부 선언
```jsx
function Modal(){

 function createRestaurantItem(event) {
  try {
    event.preventDefault();

    const restaurantData = Object.fromEntries(new FormData(this));
    restaurantDataList.addData(restaurantData);

    $(".modal")?.remove();
  } catch (e) {
    alert(e.message);
  }
}

$form.addEventListener("submit", createRestaurantItem);
}
```

외부 선언
```jsx
function Modal(){
   $form.addEventListener("submit", createRestaurantItem);
}

function createRestaurantItem(event) {
  try {
    event.preventDefault();

    const restaurantData = Object.fromEntries(new FormData(this));
    restaurantDataList.addData(restaurantData);

    $(".modal")?.remove();
  } catch (e) {
    alert(e.message);
  }
}
```

처음에는 addEventListener 내부에서 콜백 함수를 선언했지만,
콜백 함수를 함수 내부에 선언하면 해당 함수의 역할을 이해하는 데 방해가 되고,
가독성이 떨어진다고 판단하여 외부로 분리하는 방식으로 구현했습니다!!

도밥은 콜백 함수를 함수 내부에 선언하는 것과 외부에 선언하는 것 중 어떤 방식을 선호하시나요? 그리고 상황에 따라 어느 방식이 더 적절한지에 대한 의견이 궁금합니다! 😊

---

나머지 궁금한 점은 코멘트에 달았습니당~~~~~

yeongipark committed:
refactor: 모달창 폼 내부 아이템 생성하는 함수 분리

yeongipark committed:
feat: 모달 배경 누르거나 esc 누르면 닫히는 기능 추가

yeongipark committed:
refactor: 불필요한 키다운 이벤트가 유지되는 것을 막기 위해, 모달창이 닫히면 이벤트 제거하는 로직 추가

Creative-Lee commented on cypress/e2e/reload.cy.js:
step1의 요구사항을 검증하는 테스트가 되겠네요ㅎㅎ
(잘 하면 step2에서 살짝 변형해서 잘 사용해볼 수 있을것 같네요ㅋㅋ)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -23,8 +23,7 @@ export default function RestaurantModal() {
       event.preventDefault();
-      const $form = $(".restaurant-add-form");
-      const data = Object.fromEntries(new FormData($form));
+      const data = Object.fromEntries(new FormData(this));
```
this로 참조하기 보다는 비교적 더 명시적인 event.target을 사용해봅시다. 
(지금은 이벤트 핸들러에서 사용되니 this가 DOM이라는것이 보장되지만, 그렇다고 하더라도 굳이 변동적인 this를 사용할 필요는 없을 것 같습니다.)

yeongipark replied:
최근에 코어 자바스크립트로 배운 this를 활용해 보고자 이렇게 했봤습니다...!😂 

더 명시적인 방법으로 event.target이 좋아보이네요~~

yeongipark replied:
수정 커밋 : [refactor: this -> event.target으로 변경](https://github.com/woowacourse/javascript-lunch/pull/201/commits/e9f942ffb94f7653faf7d18a5d79aacf2e828aa6)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -5,81 +5,89 @@ import Input from "../util/Input.js";
+  const formItems = createFormItems();
+  $form.append(...formItems);
+  $form.appendChild(RestaurantModalButtonContainer());
+
```
[nit] 사소하지만 DOM을 담는 순서도 맥락에 맞게 정리하면 좋을 것 같습니다.

1. form 요소를 먼저 완성
2. 완성된 form과 헤더를 fragment에 담고
3. fragment return

yeongipark replied:
오 조아요!! 이렇게 하면 코드를 이해하고 구조를 파악하는데 훨씬 쉬울 거 같아요!!

yeongipark replied:
수정 커밋 : [refactor: 돔에 추가하는 순서 변경](https://github.com/woowacourse/javascript-lunch/pull/201/commits/5bdad98ee91fddd452121a951825fbbfc8aba124)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -5,81 +5,89 @@ import Input from "../util/Input.js";
+    "10분 내",
+    "15분 내",
+    "20분 내",
+    "30분 내",
```
> 저는 2번 방식을 선택했습니다! 그 이유는, 해당 데이터는 RestaurantModal 내부에서만 필요한 값이므로,
함수 외부에 선언하면 RestaurantModal이 호출되기 전부터 이미 생성되어 있어 불필요하게 메모리를 차지할 것이라고 생각했기 때문입니다!

어떤 방식이던 괜찮다고 생각합니다만, 저는 1번 방식을 선호합니다. 
말씀해주신것 처럼 함수 구현에만 집중할 수 있다는 장점이 크다고 생각해서에요.
수정을 해야할때 함수 구현내부가 아닌 파일의 상단 어딘가를 보면 된다는 사고가 가능하기도 하고요.
외부에 선언해서 메모리를 차지한다는 의견에 대해서는 '만약 함수가 여러번 실행된다면 그때마다 메모리에 할당된텐데요?' 라고 대답할 수 있을 것 같아요.(근데 사실 메모리 사용량 측면에서는 뭘 선택하던 영향이 없다고 봐도 될것 같아요 ㅋㅋ)

yeongipark replied:
저도 구현은 2번으로 했지만, 1번이 가독성 측면에서 더 좋을 거라고 생각했었어요!! 수정하겠씁니다~~

수정 커밋 :  [refactor: 상수 위치 함수 내부 -> 외부로 변경](https://github.com/woowacourse/javascript-lunch/pull/201/commits/4ee54ace5a6ea6d0b3960c9c7e26664c54600386)

Creative-Lee commented on src/components/modal/RestaurantModal.js:
```diff
@@ -5,81 +5,89 @@ import Input from "../util/Input.js";
+  } catch (e) {
+    alert(e.message);
+  }
+}
```
>addEventListener내부에 넘겨줄 콜백 함수를 분리하다가, 이를 현재 addEventListener를 호출하는 함수 내부에 함수를 선언할 지, 아니면 외부에 선언할 지 고민했습니다.

PR 본문 질문에 대해 답변 남기자면, 상황에 따라 다를 것 같은데요 뭘 선택하던 괜찮다는 주의이긴 합니다. (각 케이스에 일관성만 지켜진다면야...)

당연한 말이겠지만, 함수에서 상위 스코프의 변수등을 참조해야한다면 내부에 같이 작성해야할 것이고요.
혹은 콜백이 비교적 간단하거나, 재사용될 여지가 아예 없다거나 하면 내부에 같이 작성할 것 같아요.

그렇지 않고 지금처럼 이벤트 객체 정도만 넘겨준다거나 하는 경우엔 외부로 분리해도(바로 밑에 선언하는 정도) 괜찮을 것 같습니다. 재사용도 여지가 아예 없는것은 아니기도 하고요.

저는 거의 모든 경우에 가독성을 우선으로 생각하는데요.
지금 코드 읽는 흐름에서 
`$form.addEventListener("submit", createRestaurantItem);` 요 라인을 읽는 것 만으로도 충분히 컨텍스트가 파악되었기에 다음 라인을 읽으러 갈것 같아요. createRestaurantItem 함수의 구현부는 나중에 확인해도 될것 같기 때문에 외부에 분리되어 있어도 될 것 같다는 의견입니다~!

yeongipark committed:
feat: Select 컴포넌트 기본값과 values 속성 추가

Creative-Lee commented on src/components/util/Button.js:
공통 컴포넌트는 보통 common이라는 디렉터리에 보관하곤 합니다.
util 이라는 디렉터리명은 유틸함수를 보관하는 목적으로 자주사용되는 디렉터리명이기에 조금 어색하게 느껴지네요.


또한 지금 components 하위에 디렉터리들은 그 위계가 다소 어색하게 느껴지는데요.
우디는 어떤 기준으로 restaurant 디렉터리에 컴포넌트를 정리했는지 궁금합니다. 

제가 가진 의문은
- restaurant 디렉터리에 restaurant과 관련된것을 몰아넣을 거라면, 왜 restaurantModal은 포함되지 않았나요?
  -  header, app, item, list 어떤 공통점이 있나요?
- modal 디렉터리에는 음식점 추가 Modal과 관련된 컴포넌트만 몰아넣을거라면 디렉터리명에도 '음식점' 개념이 포함되어야 하지 않을까요?

가볍게 한번 생각해보시고 디렉터리를 정리해서 의도를 명확히 해봅시다.

yeongipark replied:
> restaurant 디렉터리에 restaurant과 관련된것을 몰아넣을 거라면, 왜 restaurantModal은 포함되지 않았나요?
header, app, item, list 어떤 공통점이 있나요?

![image](https://github.com/user-attachments/assets/07db4822-c68b-4954-8d18-ecd9640b1d72)
위 이미지에서 컴포넌트를 분리한 기준을 바탕으로 폴더 구조를 구성했습니다!

Step 1에서는 두 개의 페이지가 필요하다고 판단하여, restaurant과 modal 폴더로 나누어 관리했습니다.
- restaurant/ 폴더에는 이미지 기준 왼쪽 화면(음식점 목록 페이지)에 필요한 컴포넌트들을 저장했습니다.
- modal/ 폴더에는 이미지 기준 오른쪽 화면(음식점 추가 모달)에 필요한 컴포넌트들을 저장했습니다.

이렇게 각 페이지별로 관련된 컴포넌트들을 그룹화 했습니다!

>modal 디렉터리에는 음식점 추가 Modal과 관련된 컴포넌트만 몰아넣을거라면 디렉터리명에도 '음식점' 개념이 포함되어야 하지 않을까요?

저도 이 부분은 동의해요!!

yeongipark replied:
수정 커밋
[chore: 음식점 추가 모달에 필요한 컴포넌트들 '음식점 추가' 라는 뜻 파일명에 추가](https://github.com/woowacourse/javascript-lunch/pull/201/commits/fd3b0c362a9b10e1f0be7b854dc42e60dce58170)
[chore: Modal 컴포넌트 위치 modal 디렉토리로 이동](https://github.com/woowacourse/javascript-lunch/pull/201/commits/91313103846f2830dd373bd250d14e0636518cbb)
[chore: Modal 컴포넌트 위치 modal 디렉토리로 이동](https://github.com/woowacourse/javascript-lunch/pull/201/commits/512d3d29c992376bc558a0803007a35bd2d56205)

Creative-Lee commented on src/domain/RestaurantDataList.js:
```diff
@@ -1,10 +1,13 @@
   constructor(dataList) {
     this.#dataList = dataList.map((data) => {
       return this.createData(data);
+      a;
```
A~

yeongipark replied:
수정 커밋 : [chore: 필요한 문 삭제](https://github.com/woowacourse/javascript-lunch/pull/201/commits/7c74696038b6f6e3d664dc68c2493d76477e1181)

커밋 메시지를 반대로 작성했네요 😂

Creative-Lee reviewed (changes requested):
안녕하세요 우디~! 빠르게 반영해주셨네요 👍 

질문에 답변 드리기 전에 
>성능 최적화에 대한 고민도 있었지만, **“최적화는 문제가 발생했을 때 진행해도 늦지 않다”**는 원칙(?)을 적용했습니다.

개인적으로 훌륭한 원칙이라고 생각하고, 저도 우디와 같은 기조로 업무를 진행하고 있습니다 ㅎㅎ

다만, 지금은 학습을 하는 차원에서 
- 지금은 발생하지 않지만, 그래도 더 최적화 해본다면~
- 지금은 필요없지만, 나중에 수정되어서 필요하다면~

위와 같은 시도를 해보는것도 좋다고 생각해요 😄  (물론 시간과 정신적 여유가 허락하는 선에서만 입니다.)


---

> 0️⃣  📌 이를 위해 모달을 닫는 함수를 만들고, 해당 함수에서 modal.remove()와 함께 이벤트 제거 코드도 추가했습니다.

이벤트 제거를 위한 구현으로 봤을 때 합당해보여요. 저도 이렇게 구현했을 것 같습니다. 


> 1️⃣ 상수화 위치

코드에 코멘트 남겨두었습니다 ㅎㅎ

> 2️⃣ 선언형 방식

우디의 코드에서 찾기는 조금 어려운 것 같아서, 지난 미션들에서 명령형/선언형 관련 질문에 항상 첨부했던 [LMS 학습자료 링크](https://techcourse.woowahan.com/s/DL2M0i9l/ls/RY6ZuXMo) 걸어두겠습니다~! 

어떤 코드가 명령형이나 선언형에 가까운지는 코드를 많이 읽어보면 자연스럽게 체화되는 부분인것 같기도 합니다.
어떤것을 얻기 위해 그 방법을 상세하게 명시하는가 혹은 그저 원하는 것을 말하는가 를 위주로 생각해봐도 좋을 것 같아요.

> 3️⃣ 콜백 함수 선언 위치

이 내용도 1번 질문과 비슷한 맥락이라고 생각합니다. 
코멘트에 같이 남겨두었어요~!

코멘트 가볍게 한번 확인해보시죠~! ㅎㅎ

yeongipark committed:
chore: 공통 컴포넌트 보관하는 폴더명 common으로 변경

yeongipark committed:
chore: 음식점 추가 모달에 필요한 컴포넌트들 '음식점 추가' 라는 뜻 파일명에 추가

yeongipark committed:
chore: Modal 컴포넌트 위치 modal 디렉토리로 이동

yeongipark committed:
chore: Modal 컴포넌트 위치 modal 디렉토리로 이동

yeongipark committed:
chore: RestaurantAddModal에 필요한 css 하나의 파일로 수정

yeongipark committed:
chore: 필요한 문 삭제

yeongipark committed:
test: 각 테스트는 독립적이므로, 모달창이 열렸는지 확인하는 코드 추가

yeongipark committed:
refactor: this -> event.target으로 변경

yeongipark committed:
refactor: 돔에 추가하는 순서 변경

yeongipark committed:
refactor: 상수 위치 함수 내부 -> 외부로 변경

yeongipark commented:
@Creative-Lee 
안녕하세요 도밥!! 😊

개인적인 일이 있어서 피드백 반영이 예상보다 오래 걸렸네요… 🥲
매번 제 질문에 상세하게 답변해주시고, 꼼꼼하게 코드 리뷰해주셔서 정말 많이 배우고 생각할 기회가 된 것 같아요!
도밥 같은 리뷰어를 만날 수 있어서 너무 좋습니다! 🙌 감사합니다! 🙏✨

---

>옵저버 패턴외에 고려했던 다른 방법을 설명해주시겠어요?

사실, 옵저버 패턴 외에 다른 방법을 고려하지 않았습니다..!

처음에는 Restaurant 컴포넌트에서 초기 화면에 필요한 컴포넌트들을 body에 추가하는 구조를 유지하고 싶었습니다.
여러 컴포넌트에서 개별적으로 body에 추가하는 방식이 되면, 어느 시점에 추가되는지 추적하기 어려워질 것이라고 생각했기 때문입니다!

따라서, 화면 전체를 다시 그리지 않고, 필요한 부분만 리렌더링하는 방법을 고민했습니다.
처음에는 Restaurant 컴포넌트를 전부 지우고, 상태가 변경될 때 다시 추가하는 방식으로 구현했습니다.
그러나, 이 방식은 굳이 다시 렌더링할 필요가 없는 RestaurantHeader까지 삭제되는 문제가 발생했습니다!

이런 고민을 하던 중 옵저버 패턴을 알게 되었고, 옵저버 패턴을 적용하면 상태가 변경될 때, 해당 상태를 구독하는 곳에서만 리렌더링이 발생할 수 있습니다.
즉, 불필요한 컴포넌트 리렌더링 없이 RestaurantListContainer만 업데이트할 수 있다는 점에서 적절한 해결책이라고 판단했습니다!!

그래서 다른 방법을 고려하지 않고 바로 옵저버 패턴을 선택하여 구현했습니다!


### 옵저버 패턴 단점

처음에는 옵저버 패턴을 적용하고 나서 “단점이 없는 것 같은데?” 라고 생각했습니다!
그러나, 모든 기술에는 트레이드오프가 있듯이, 옵저버 패턴도 단점이 존재할 것이라 생각합니다!

제가 직접 체감하지는 못했지만, 이론적으로 생각해본 옵저버 패턴의 단점은 다음과 같습니다!

1️⃣ 구독자가 많아지면 불필요한 구독자가 생길 가능성이 있음.

2️⃣ 불필요하게 상태가 변경될 가능성이 있음

3️⃣ 디버깅이 어려워질 수 있음

하지만 현재 규모에서는 옵저버 패턴의 단점이 크게 체감되지 않아요...!

Creative-Lee commented on index.html:
```diff
@@ -2,12 +2,163 @@
 <html lang="en">
```
[nit] step2 시작하면서 한국으로 바꿔줍시다~!

Creative-Lee reviewed (approved):
나머지 코멘트도 적절하게 반영해주셨네요 ㅎㅎ


저도 잘 모르지만, 옵저버 패턴에 대한 생각을 적어주신걸 보고 문득
'당장 구독자는 1명임에도 랜더링 최적화를 위해 보일러 코드(구독자 필드, 구독, notify등등)를 작성하면서까지 도입할만 한가? 오버엔지니어링은 아닐까?' 라는 고민이 생기네요~!
다른 방법은 생각하지 않으셨다고 하니, step2 진행하고 남는 시간에 다른 방법은 뭐가 있을지? 조금 고민해보셔도 좋겠습니다 ㅎㅎ(바꾸라는 말 아닙니다!!!)

별개로 배포된 앱 눌러보면서, 음식점 추가할 때 `1글자 음식점`은 조금 억울하겠다는 생각이 들었네요ㅎㅎ

고생하셨습니다 우디~! step2에서 뵙겠습니다 👋
